[DOC] First draft release notes (#2022)
grafana/tempo: docs/sources/release-notes/v2-0.md

- The metrics-generator now handles collisions between user-defined and default dimensions [PR 1794](https://github.com/grafana/tempo/pull/1794).
- Metrics-generator now makes intrinsic dimensions configurable and disables `status_message` by default [PR 1960](https://github.com/grafana/tempo/pull/1960).
- Span sizes can be exposed as a metric [PR 1662](https://github.com/grafana/tempo/pull/1662).
- Older spans are now filtered out before metrics are aggregated [PR 1612](https://github.com/grafana/tempo/pull/1612).
- Label names and values that exceed a configurable length are truncated [PR 1897](https://github.com/grafana/tempo/pull/1897).

In addition to the updates to search and metrics-generators, ingesters now have zone awareness replication.

```yaml
# use the following fields in _config field of jsonnet config, to enable zone aware ingesters.
    multi_zone_ingester_enabled: false,
    multi_zone_ingester_migration_enabled: false,
    multi_zone_ingester_replicas: 0,
    multi_zone_ingester_max_unavailable: 25,
```

## Upgrade considerations

When upgrading to Tempo 2.0, be aware of these breaking changes.

### Defaults updated

> The following configuration parameters have been updated with new default values. These defaults were changed to work better with the new Parquet block format. [PR 1978](https://github.com/grafana/tempo/pull/1978)

Author: ```suggestion
The following configuration parameters have been updated with new default values. These defaults were changed to work better with the new Parquet block format. [PR 1978](https://github.com/grafana/tempo/pull/1978)

If you used the Parquet format with Tempo 1.5, we recommend updating your configuration to use these default values. 
```

### TraceQL query editor in Grafana

To make it easier to run and visualize the results of TraceQL queries, we’ve added a TraceQL query editor to Grafana’s Tempo data source. The editor is available in Grafana’s Explore interface.

<p align="center"><img src="../traceql/assets/query-editor-results-span.png" alt="Query editor showing span results" /></p>

The query editor helps you learn TraceQL by offering autocomplete.

<p align="center"><img src="../traceql/assets/query-editor-auto-complete.png" alt="Query editor showing the auto-complete feature" /></p>

For more information, refer to the [TraceQL query editor documentation](https://grafana.com/docs/tempo/latest/traceql/query-editor.md).

### Apache Parquet block format is the new default

<!-- JOE -- NEED TEST RESULTS -->

The experimental Apache Parquet columnar block format introduced in Tempo v1.5 has been promoted to stable and is now the default trace storage for Tempo 2.0. This format enables increased speed and efficiency for Tempo search.

> In our own environments, we’re seeing that it allows us to scan trace data at rates of 1 TB/s (with some bursts as high as 4 TB/s), a massive improvement over the **XXX** rates that were more typical with our previous storage format. Switching to the Parquet block format also allows users to leverage the large ecosystem of existing Parquet tools and libraries to process and transform their trace data.

Author: @joe-elliott  Did you see what the rates would be in your research? Need to replace the **XXX** with the correct info.

Author: I updated the sentence to information that we had in the 1.5 blog post for data for Parquet. we can update if we get more recent info, but at least this is still an impressive number.

> In our own environments, we’re seeing that it allows us to scan trace data at rates of 1 TB/s (with some bursts as high as 4 TB/s), a massive improvement over the **XXX** rates that were more typical with our previous storage format. Switching to the Parquet block format also allows users to leverage the large ecosystem of existing Parquet tools and libraries to process and transform their trace data.

Author: ```suggestion
In our own environments, we’re seeing that it allows us to scan trace data at rates of 1 TB/s (with some bursts as high as 4 TB/s), a massive 1.5x improvement over the rates that were more typical with our previous storage format. Switching to the Parquet block format also allows users to leverage the large ecosystem of existing Parquet tools and libraries to process and transform their trace data.
```

Tempo search will return results from both the legacy and Parquet block formats. However, TraceQL will only search Parquet. Search on legacy blocks is deprecated and will be removed in Tempo 2.1.
For more information, refer to the Parquet documentation.

Refer to the [Parquet documentation]({{< relref "../configuration/parquet/" >}}).

### Search and metrics-generator enabled by default

Search and [metrics-generator]({{< relref "../metrics-generator/" >}}) are now enabled by default in Tempo 2.0, and their corresponding configuration parameters (search_enabled and metrics_generator_enabled ) have been removed).

> We’ve been excited by the feedback we’ve gotten on both features and are satisfied with their stability so we want all newcomers to Tempo to have access to them from day 1.

Author: ```suggestion
We’ve been excited by the feedback we’ve gotten on both features and are satisfied with their stability so we want all newcomers to Tempo to have access to them from day 1.

The metrics-generator component is now enabled by default, however, you still still need to tell it to derive metrics for a tenant by using tenant-specific configurations. For more information, refer to the [metrics-generator documentation]{{< relref "../metrics-generator" >}}). 
```

> >**NOTE:** The metrics-generator is not turned on by default for Grafana Cloud customers due to the potential cost implications.

Author: Will Cloud customers still have to contact support to enable metrics-generator? Or do they just enable in a configuration file some where?

> >**NOTE:** The metrics-generator is not turned on by default for Grafana Cloud customers due to the potential cost implications.

Author: ```suggestion
>**NOTE:** Grafana Cloud customer still need to contact Grafana Support to enable metrics generation. 
```

### Other upgrade considerations

[PR 2004](https://github.com/grafana/tempo/pull/2004) The  `search_enabled` and `metrics_generator_enabled` configuration parameters have been removed. Both default to true.

[PR 1879](https://github.com/grafana/tempo/pull/1879). Use snake case on Azure Storage config. Here is an example of using snake case on Azure Storage config:

```yaml
# config.yaml
storage:
  trace:
    azure:
      storage_account_name:
      storage_account_key:
      container_name:
```

[PR 1678](https://github.com/grafana/tempo/pull/1678). Parquet is the new default block version. To continue using the v2 backend set, see the [Parquet configuration documentation](https://grafana.com/docs/tempo/latest/configuration/parquet/).

[PR 1810](https://github.com/grafana/tempo/pull/1810). Delete `TempoRequestErrors` alert from mixin. Any Jsonnet users relying on this alert should copy this into their own environment.

[PR 1794](https://github.com/grafana/tempo/pull/1794). metrics-generator: Handle collisions between user defined and default dimensions. Custom dimensions colliding with intrinsic dimensions will be prefixed with `__`.

> [PR 1874](https://github.com/grafana/tempo/pull/1874). Add TLS support to the vulture. Internal types are updated to use `scope` instead of `instrumentation_library`. This is a breaking change in trace by ID queries if JSON is requested.

Author: ```suggestion
[PR 1754](https://github.com/grafana/tempo/pull/1754). Add TLS support to the vulture. Internal types are updated to use `scope` instead of `instrumentation_library`. This is a breaking change in trace by ID queries if JSON is requested.
```

## Features and enhancements

Some of the most important features and enhancements in Tempo 2.0 are highlighted below.

### TraceQL, the trace query language

Inspired by PromQL and LogQL, TraceQL is a query language designed for selecting traces in Tempo. TraceQL makes it possible to find traces that otherwise would have been difficult or even impossible to identify with Tempo's existing search. It also makes it possible for you to understand the broad context for an event sequence.
To understand more about why we built TraceQL and the use cases we think it unlocks, see [our blog post](https://grafana.com/blog/2022/11/30/traceql-a-first-of-its-kind-query-language-to-accelerate-trace-analysis-in-tempo-2.0/).

With this release, we’ve implemented a subset of the full TraceQL language we envision. Users can select traces based on the following:

- Span and resource attributes, timing, and duration
- Basic aggregates: `count()` and `avg()`

To learn more about the TraceQL syntax, see the [TraceQL documentation]({{< relref "../traceql" >}}).
For information on how we plan to extend the TraceQL language, see [future work]({{< relref "../traceql/architecture" >}}).

> You can run a TraceQL query either by issuing it to Tempo’s [`search_tag_values_v2` API endpoint](https://grafana.com/docs/tempo/latest/api_docs/#search-tag-values-v2), or, for those using Tempo in conjunction with Grafana, by using Grafana’s TraceQL query editor, described below.

Author: ```suggestion
You can use the new v2 API endpoints to access Tempo . For example, uou can run a TraceQL query either by issuing it to Tempo’s [`search_tag_values_v2` API endpoint](https://grafana.com/docs/tempo/latest/api_docs/#search-tag-values-v2), or, for those using Tempo in conjunction with Grafana, by using Grafana’s TraceQL query editor, described below.
```

> You can run a TraceQL query either by issuing it to Tempo’s [`search_tag_values_v2` API endpoint](https://grafana.com/docs/tempo/latest/api_docs/#search-tag-values-v2), or, for those using Tempo in conjunction with Grafana, by using Grafana’s TraceQL query editor, described below.

Author: Joe updated the API page, so I've updated the link in the text.

### TraceQL query editor in Grafana

To make it easier to run and visualize the results of TraceQL queries, we’ve added a TraceQL query editor to Grafana’s Tempo data source. The editor is available in Grafana’s Explore interface.

<p align="center"><img src="../traceql/assets/query-editor-results-span.png" alt="Query editor showing span results" /></p>

The query editor helps you learn TraceQL by offering autocomplete.

<p align="center"><img src="../traceql/assets/query-editor-auto-complete.png" alt="Query editor showing the auto-complete feature" /></p>

For more information, refer to the [TraceQL query editor documentation](https://grafana.com/docs/tempo/latest/traceql/query-editor.md).

### Apache Parquet block format is the new default

The experimental Apache Parquet columnar block format introduced in Tempo v1.5 has been promoted to stable and is now the default trace storage for Tempo 2.0. This format enables increased speed and efficiency for Tempo search.

> In our own environments, we’re seeing that it allows us to scan trace data at rates of 1 TB/s (with some bursts as high as 4 TB/s). Previously, we capped out around 40 to 50 GB/s searched on the old format. With Parquet, we are currently seeing closer to 300 GB/s searched on common queries with fewer resources used. Switching to the Parquet block format also allows users to leverage the large ecosystem of existing Parquet tools and libraries to process and transform their trace data.

Author: The content that I added was based upon the Tempo 1.5 release blog post that @joe-elliott wrote. @joe-elliott  Are these values okay to use in the 2.0 release notes?

> In our own environments, we’re seeing that it allows us to scan trace data at rates of 1 TB/s (with some bursts as high as 4 TB/s). Previously, we capped out around 40 to 50 GB/s searched on the old format. With Parquet, we are currently seeing closer to 300 GB/s searched on common queries with fewer resources used. Switching to the Parquet block format also allows users to leverage the large ecosystem of existing Parquet tools and libraries to process and transform their trace data.

Author: I'm updating based upon Joe's returned numbers.

For more information, refer to the Parquet documentation.

Refer to the [Parquet documentation]({{< relref "../configuration/parquet/" >}}).

### Search and metrics-generator enabled by default

Search and [metrics-generator]({{< relref "../metrics-generator/" >}}) are now enabled by default in Tempo 2.0, and their corresponding configuration parameters (search_enabled and metrics_generator_enabled ) have been removed).

We’ve been excited by the feedback we’ve gotten on both features and are satisfied with their stability so we want all newcomers to Tempo to have access to them from day 1.

The metrics-generator component is now enabled by default, however, you still still need to tell it to derive metrics for a tenant by using tenant-specific configurations. For more information, refer to the [metrics-generator configuration documentation]{{< relref "../configuration" >}}). 

>**NOTE:** Grafana Cloud customer need to contact Grafana Support to enable metrics generation. 

These enhancements have been added to the metrics-generator:

> - The metrics-generator now handles collisions between user-defined and default dimensions [PR 1794](https://github.com/grafana/tempo/pull/1794).

Reviewer: ```suggestion
- The metrics-generator now handles collisions between user-defined and default dimensions [PR 1794](https://github.com/grafana/tempo/pull/1794). User-defined dimensions colliding with default dimensions will be prefixed with `__`.
```

### Defaults updated

The following configuration parameters have been updated with new default values. These defaults were changed to work better with the new Parquet block format. [PR 1978](https://github.com/grafana/tempo/pull/1978)

If you used the Parquet format with Tempo 1.5, consider updating your configuration to use these default values. 

```yaml
query_frontend:
    max_oustanding_per_tenant: 2000
    search:
        concurrent_jobs: 1000
        target_bytes_per_job: 104857600
        max_duration: 168h
        query_ingesters_until: 30m
    trace_by_id:
        query_shards: 50
querier:
    max_concurrent_queries: 20
    search:
        prefer_self: 10
ingester:
    concurrent_flushes: 4
    max_block_duration: 30m
    max_block_bytes: 524288000
storage:
    trace:
        pool:
            max_workers: 400
            queue_depth: 20000
        search:
            read_buffer_count: 32
            read_buffer_size_bytes: 1048576

```

### Removed and renamed configuration parameters

The following tables describe the parameters that have been removed or renamed. (PR#1978)

#### Removed

| Parameter | Comments |
| :-- | :-- |

> | query_frontend:<br/>  query_shards: |   |

Reviewer: ```suggestion
```

> | querier:<br/>  query_timeout: | Use `trace_by_id.query_shards` |

Reviewer: ```suggestion
```

| ingester:<br/>  use_flatbuffer_search: | Automatically set based on block type |

> | storage:<br/>  wal:<br/>    version: | Removed and pinned to `block.version` |

Reviewer: ```suggestion
| search_enabled | Removed and set to true |
| metrics_generator_enabled | Removed and set to true |
| storage:<br/>  wal:<br/>    version: | Removed and pinned to `block.version` |
```

#### Renamed

| Parameter | Comments |

> | --- | --- |

Reviewer: ```suggestion
| --- | --- |
| query_frontend:<br/>  query_shards: | Renamed `query_frontend.trace_by_id.query_shards` |
| querier:<br/>  query_timeout: | Renamed `querier.trace_by_id.query_timeout` |
```

| `compactor` section |  |
| compaction:<br/>  chunk_size_bytes: | Renamed to `v2_in_buffer_bytes` |
| compaction:<br/>  flush_size_bytes: | Renamed to `v2_out_buffer_bytes` |
| compaction:<br/>  iterator_buffer_size: | Renamed to `v2_prefetch_traces_count` |
| `storage` section |  |
| wal:<br/>  encoding: | Renamed to `v2_encoding` |
| block:<br/>  index_downsample_bytes: | Renamed to `v2_index_downsample_bytes` |
| block:<br/>  index_page_size_bytes: | Renamed to `v2_index_page_size_bytes` |
| block:<br/>  encoding: | Renamed to `v2_encoding` |
| block:<br/>  row_group_size_bytes: | Renamed to `parquet_row_group_size_bytes` |

### Other upgrade considerations

> [PR 2004](https://github.com/grafana/tempo/pull/2004) The  `search_enabled` and `metrics_generator_enabled` configuration parameters have been removed. Both default to true.

Reviewer: ```suggestion
```

[PR 1879](https://github.com/grafana/tempo/pull/1879). Use snake case on Azure Storage config. Here is an example of using snake case on Azure Storage config:

```yaml
# config.yaml
storage:
  trace:
    azure:
      storage_account_name:
      storage_account_key:
      container_name:
```

> [PR 1678](https://github.com/grafana/tempo/pull/1678). Parquet is the new default block version. To continue using the v2 backend set, see the [Parquet configuration documentation](https://grafana.com/docs/tempo/latest/configuration/parquet/).

Reviewer: What is the "v2 backend set"? Is that what the old block version was called?

Author: It refers to using the v2 block format, so I've updated the content. While creating the Upgrade to Tempo doc with Marty, we rewrote that line to say block format instead. I've made the same change here. 
```suggestion
[PR 1678](https://github.com/grafana/tempo/pull/1678). Parquet is the new default block version. To continue using the v2 block format, see the [Parquet configuration documentation](https://grafana.com/docs/tempo/latest/configuration/parquet/).
```

> [PR 1810](https://github.com/grafana/tempo/pull/1810). Delete `TempoRequestErrors` alert from mixin. Any Jsonnet users relying on this alert should copy this into their own environment.

Reviewer: ```suggestion
[PR 1810](https://github.com/grafana/tempo/pull/1810). We've deleted the `TempoRequestErrors` alert from mixin. Any Jsonnet users relying on this alert should copy this into their own environment.
```

> [PR 1794](https://github.com/grafana/tempo/pull/1794). metrics-generator: Handle collisions between user defined and default dimensions. Custom dimensions colliding with intrinsic dimensions will be prefixed with `__`.

Reviewer: ```suggestion
```
We already mentioned as an enhancement.

> [PR 1754](https://github.com/grafana/tempo/pull/1754). Add TLS support to the vulture. Internal types are updated to use `scope` instead of `instrumentation_library`. This is a breaking change in trace by ID queries if JSON is requested.

Reviewer: ```suggestion
[PR 1754](https://github.com/grafana/tempo/pull/1754). We've added TLS support to the vulture. Internal types are updated to use `scope` instead of `instrumentation_library`. This is a breaking change in "trace by ID" type queries if JSON is requested.
```

## Features and enhancements

Some of the most important features and enhancements in Tempo 2.0 are highlighted below.

### TraceQL, the trace query language

Inspired by PromQL and LogQL, TraceQL is a query language designed for selecting traces in Tempo. TraceQL makes it possible to find traces that otherwise would have been difficult or even impossible to identify with Tempo's existing search. It also makes it possible for you to understand the broad context for an event sequence.
To understand more about why we built TraceQL and the use cases we think it unlocks, see [our blog post](https://grafana.com/blog/2022/11/30/traceql-a-first-of-its-kind-query-language-to-accelerate-trace-analysis-in-tempo-2.0/).

With this release, we’ve implemented a subset of the full TraceQL language we envision. Users can select traces based on the following:

- Span and resource attributes, timing, and duration
- Basic aggregates: `count()` and `avg()`

To learn more about the TraceQL syntax, see the [TraceQL documentation]({{< relref "../traceql" >}}).
For information on how we plan to extend the TraceQL language, see [future work]({{< relref "../traceql/architecture" >}}).

> You can run a TraceQL query either by issuing it to Tempo’s [`search` API endpoint]({{< relref "../api_docs/#search" >}}), or, for those using Tempo in conjunction with Grafana, by using Grafana’s TraceQL query editor, described below.

Author: Changed the endpoint to the `search` area of the API doc.

> You can run a TraceQL query either by issuing it to Tempo’s [`search` API endpoint]({{< relref "../api_docs/#search" >}}), or, for those using Tempo in conjunction with Grafana, by using Grafana’s TraceQL query editor, described below.

Author: ```suggestion
You can run a TraceQL query either by issuing it to Tempo’s `q` parameter of the [`search` API endpoint]({{< relref "../api_docs/#search" >}}), or, for those using Tempo in conjunction with Grafana, by using Grafana’s TraceQL query editor, described below.
```

The query editor helps you learn TraceQL by offering autocomplete.

<p align="center"><img src="../../traceql/assets/query-editor-auto-complete.png" alt="Query editor showing the auto-complete feature" /></p>

For more information, refer to the [TraceQL query editor documentation]({{< relref "../traceql/query-editor" >}}).

### Apache Parquet block format is the new default

The experimental Apache Parquet columnar block format introduced in Tempo v1.5 has been promoted to stable and is now the default trace storage for Tempo 2.0. We're excited to switch to it because of the increased speed and efficiency it provides for search.

With our previous block format, we'd seen our ability to search trace data cap out at a rate of ~40-50 GB per second. Since switching to Parquet, we're now hitting search speeds of 300 GB/s on common queries and doing so with less compute. If we define the bytes processed as the total trace bytes that are scanned, then this example query resulted in ~1.2 TB/s:

```yaml
{ resource.cluster ="foo" }
```

and this query resulted in ~550 GB/s:

```yaml
{ span.foo ="bar" }
```

> 

Author: This is the section that is being reviewed for the numbers.
